### PR TITLE
name_converter functionality added to SGTK templates

### PR DESCRIPTION
**Pull request by garvitverma**
This will allow you to convert sg_client_name field to code field, for any entity type, and vice versa.

### CODE REVIEW

File: core/hooks/name_converter.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+"""
+Hook which chooses an environment file to use based on the current context.
+
+"""
+import os
+import copy
+
+import sgtk
+from tank.util.shotgun import get_sg_connection
```
Comment (mbrainerd) [inline core/hooks/name_converter.py]: This line should be removed.

File: core/hooks/name_converter.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        """
+        Test if a value is valid for this key
+
+        :param value: Value to test
+        :param validate_transforms: Bool to enable validation of transforms,
+                                    such as a subset calculation
+        :returns: Bool
+        """
+        # to avoid the validation to fail, if the value is not defined in choices!!
+        return True
```
Comment (mbrainerd) [inline core/hooks/name_converter.py]: Hrm....this seems bad. Can we talk about why this is necessary?

File: core/templates.yml [hunk 3]
```diff
@@ -60,12 +60,63 @@ keys:
         type: str
         shotgun_entity_type: HumanUser
         shotgun_field_name: login
+
     custom_key_example:
         type: str
         validate_hook: 'templatekey_custom'
```
Comment (mbrainerd) [inline core/templates.yml]: You could also *just* define the value_from_str_hook and leave validate_hook and str_from_value_hook blank and the code will fallback on the default functionality.

Comment (garvitverma) [reply core/templates.yml]: Will do that in ingest.yml, although the hook is also calling the parent.
We can define the keys in ingest.yml. And maybe basic examples of these templates. But these will mostly be sourced from show's preference files for sgtk_config. As per the client tech specs.

File: core/hooks/name_converter.py (hunk 2 above)
Comment (garvitverma) [reply core/hooks/name_converter.py]: Let me update the stack trace on the ticket, which was the reason I added this hack.

File: core/templates/ingest.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+keys:
+    # examples for name_converter hook!
+    client_output:
```
[the remark this answers is not in the document]
Comment (garvitverma) [reply core/templates/ingest.yml]: Yes, I'm going to update the keys, when I actually start using them in templates as keys. Right now these are just dummy keys, That I was using for testing the hooks. I will keep that in mind while creating the new keys for "Ingest".

File: core/templates/ingest.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        choices:
+            EDIT: underscore_to_camelcase
+        default: EDIT
+        validate_hook: 'name_converter'
+        value_from_str_hook: 'name_converter'
+
+paths:
+
+    # sample template for name_converter
+    client_shot_publish_cache:              './{client_sequence}/{Shot}/SHARED/ANIM/{client_step}/@client_shot_version_name.abc'
```
[the remark this answers is not in the document]
Comment (garvitverma) [reply core/templates/ingest.yml]: Mike, This was just a test template. I'm yet to create templates that will be used when integrating it in Element Ingest. For now please ignore the contents of this file, and take it more as placeholder for the upcoming templates. I will keep the suggestions in mind.

File: core/templates/ingest.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
+keys:
+    # examples for name_converter hook!
+    client_output:
+        type: str
+        filter_by: alphanumeric
+        value_from_str_hook: 'name_converter'
+        edit_type: replace
```
[the remark this answers is not in the document]
Comment (garvitverma) [reply core/templates/ingest.yml]: Added this functionality, to differentiate between what edit_type will use what edits added another hierarchy of that in edits, you can see that in replace type edit.

**Pull request merged**